README.md
# arp-lookup

[![checks](https://github.com/justintaddei/arp-lookup/actions/workflows/tests.yml/badge.svg?branch=master)](https://github.com/justintaddei/arp-lookup/actions/workflows/tests.yml)
![](https://img.shields.io/github/issues-raw/justintaddei/arp-lookup.svg?style=flat)
![](https://img.shields.io/npm/v/@network-utils/arp-lookup.svg?style=flat)
![](https://img.shields.io/npm/dt/@network-utils/arp-lookup.svg?style=flat)
![](https://img.shields.io/npm/l/@network-utils/arp-lookup.svg?style=flat)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat)](https://github.com/prettier/prettier)
![](https://img.shields.io/badge/language-typescript-blue.svg?style=flat)

## About

`arp-lookup` is a simple ARP utility to map an IP address to a MAC address and vice versa.

> Vendor information has been removed from v2 in favor of [@network-utils/vender-lookup](https://www.npmjs.com/package/@network-utils/vendor-lookup)  
> Please see the [CHANGELOG.md](CHANGELOG.md) for the full list of breaking changes

## Installation

```bash
$ npm install @network-utils/arp-lookup
```

## Usage

### Example

```typescript
import arp from '@network-utils/arp-lookup'
// Or
import {toMAC, toIP, ...} from '@network-utils/arp-lookup'

// Retrieve the corresponding IP address for a given MAC address
await arp.toIP('04-A1-51-1B-12-92') // Or '04:a1:51:1b:12:92' (any valid MAC format)
// Result: "192.168.2.47"

// Retrieve the corresponding MAC address for a given IP address
await arp.toMAC('192.168.2.47')
// Result: "04:a1:51:1b:12:92" 👈🏼 All MAC addresses are normalized to this format

arp.isMAC('04-a1:51:1B-12-92') // true
arp.isMAC('not:a:mac') // false

arp.isIP('192.168.2.47') // true
arp.isIP('not.an.ip') // false

// Note: Unavailable on darwin based systems.
await arp.isType('dynamic', '192.168.2.47') // true
await arp.isType('dynamic', '04:a1:51:1b:12:92') // true
await arp.isType('static', '192.168.2.255') // true
await arp.isType('static', 'ff:ff:ff:ff:ff:ff') // true
await arp.isType('undefined', '0.0.0.0') // true

// Note: `type` property is always set to `"unknown"` on darwin systems
await arp.getTable()
// Result:
[
    { ip: '192.168.137.255', mac: 'ff:ff:ff:ff:ff:ff', type: 'static' },
    { ip: '224.0.0.22', mac: '01:00:5e:00:00:16', type: 'static' },
    { ip: '224.0.0.251', mac: '01:00:5e:00:00:fb', type: 'static' },
    { ip: '224.0.0.252', mac: '01:00:5e:00:00:fc', type: 'static' },
    { ip: '239.255.255.250', mac: '01:00:5e:7f:ff:fa', type: 'static' },
    { ip: '192.168.2.1', mac: '04:a1:51:1b:12:92', type: 'dynamic' },
    { ip: '192.168.2.3', mac: '1a:b1:61:2f:14:72', type: 'dynamic' },
    { ip: '192.168.2.255', mac: 'ff:ff:ff:ff:ff:ff', type: 'static' },
    { ip: '224.0.0.2', mac: '01:00:5e:00:00:02', type: 'static' },
    ...
]


// Note: `type` property is always set to `"unknown"` on Unix systems
await arp.fromPrefix('01:00:5e')
// Result:
[
    { ip: '224.0.0.22', mac: '01:00:5e:00:00:16', type: 'static' },
    { ip: '224.0.0.251', mac: '01:00:5e:00:00:fb', type: 'static' },
    { ip: '224.0.0.252', mac: '01:00:5e:00:00:fc', type: 'static' },
    { ip: '239.255.255.250', mac: '01:00:5e:7f:ff:fa', type: 'static' },
    { ip: '224.0.0.2', mac: '01:00:5e:00:00:02', type: 'static' },
    ...
]
```

---

### `getTable(): Promise<IArpTable>`

Returns a promise containing the parsed output of `$ arp -a` with the addition of a `vendor` field.  
**Note that the `type` property is always set to `"unknown"` on Unix systems**

---

### `toMAC(ip: string): Promise<string | null>`

Returns a promise containing the MAC that relates to `ip` or `null` if a match couldn't be made.  
**Throws** an `"Invalid IP"` error if `ip` is not a valid IP address

---

### `toIP(mac: string): Promise<string | null>`

Returns a promise containing the IP that relates to `mac` or `null` if a match couldn't be made.  
**Throws** an `"Invalid MAC"` error if `mac` is not a valid MAC address

---

### `fromPrefix(prefix: string): Promise<IArpTableRow[]>`

Returns any devices on the network with the specified MAC prefix, or an empty array if none exist.  
**Throws** an `"Invalid Prefix"` error if `prefix` is not a valid MAC address prefix


---

### `getType(address: string): Promise<'static' | 'dynamic' | 'unknown' | 'undefined'>`

- `address` can be any valid IP or MAC address

Returns a promise containing a string which indicates the record `type`.  
**This method is useless on Unix based systems because `$ arp -a` doesn't return the type for an address**  
**Throws** an `"Invalid address"` error if `address` is not a valid IP or MAC address

---

### `isType(type: 'static' | 'dynamic' | 'unknown' | 'undefined', address: string): Promise<boolean>`

- `address` can be any valid IP or MAC address

Returns a promise containing a boolean which indicates the record for `address` is `type`.  
Pass `type = "undefined"` to determine if a record for `address` exists or not.  
**This method is useless on Unix based systems because `$ arp -a` doesn't return the type for an address**  
**Throws** an `"Invalid address"` error if `address` is not a valid IP or MAC address

---

### `isMAC(mac: string): boolean`

Checks if a MAC address is valid

---

### `isPrefix(prefix: string): boolean`

Checks if a MAC address prefix is valid

---

### `isIP(ip: string): boolean`

Checks if an IP address is valid

---

### `IArpTableRow`

```typescript
IArpTableRow {
  ip: string
  mac: string
  type: 'static' | 'dynamic' | 'unknown'
}
```

---

### `IArpTable`

An array of `IArpTableRow`'s.
```typescript
type IArpTable = IArpTableRow[]
```

---

## Testing

```bash
$ git clone https://github.com/justintaddei/arp-lookup.git
$ cd arp-lookup
$ npm install
$ npm test
```

## License

MIT


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.1.0](https://github.com/justintaddei/arp-lookup/compare/v2.0.2...v2.1.0) (2023-09-27)


### Features

* added getType function ([64bdab8](https://github.com/justintaddei/arp-lookup/commit/64bdab8eb194b6015bfc9c7a1bb0630212eed4dd))


### Bug Fixes

* **deps:** bump actions/cache from 3.3.1 to 3.3.2 ([0ea973d](https://github.com/justintaddei/arp-lookup/commit/0ea973d3ed163122029a7de99ad9c350bb76f339))
* **deps:** bump actions/setup-node from 3.6.0 to 3.7.0 ([5c68547](https://github.com/justintaddei/arp-lookup/commit/5c6854739b0606c309abc55631f3207e105cf5fd))
* **deps:** bump actions/setup-node from 3.7.0 to 3.8.0 ([29449e0](https://github.com/justintaddei/arp-lookup/commit/29449e00cb8e2b5d4db80c2481e83399e157c9b1))
* **deps:** bump actions/setup-node from 3.8.0 to 3.8.1 ([84e0023](https://github.com/justintaddei/arp-lookup/commit/84e00237fa38df8f517e015400f8d7cb73790e98))
* **deps:** bump dependabot/fetch-metadata from 1.4.0 to 1.5.0 ([5f63861](https://github.com/justintaddei/arp-lookup/commit/5f63861b1d5e544420a83dbdc7a02f2cdb7c31db))
* **deps:** bump dependabot/fetch-metadata from 1.5.0 to 1.5.1 ([29cb44f](https://github.com/justintaddei/arp-lookup/commit/29cb44fdf184df74bf3a9d9562bbdc23cb5b26ba))
* **deps:** bump dependabot/fetch-metadata from 1.5.1 to 1.6.0 ([cd6bff2](https://github.com/justintaddei/arp-lookup/commit/cd6bff2717afd7f273b03275b6b2ddb759bb69ee))
* use '-an' instead of '-a' on Darwin/Linux ([4408591](https://github.com/justintaddei/arp-lookup/commit/4408591e8a94e1da229cea555b7b5155e2566b82))
* warnings ([cf238d4](https://github.com/justintaddei/arp-lookup/commit/cf238d4f0bf872c01d71110c2ac5b5c3530cdf49))

### [2.0.2](https://github.com/justintaddei/arp-lookup/compare/v2.0.0...v2.0.2) (2023-05-09)


### Bug Fixes

* **deps:** bump actions/cache from 2.1.7 to 3.0.1 ([80da0e7](https://github.com/justintaddei/arp-lookup/commit/80da0e74702319af309cd55aedcab0b7f83cdb39))
* **deps:** bump actions/cache from 3.0.1 to 3.0.3 ([f41b01d](https://github.com/justintaddei/arp-lookup/commit/f41b01d657dd1f3510ae6bd2f026ceb8a199f561))
* **deps:** bump actions/cache from 3.0.3 to 3.0.4 ([2892f90](https://github.com/justintaddei/arp-lookup/commit/2892f905c010feccf7c0939ec25fbdf1f31f9439))
* **deps:** bump actions/cache from 3.0.4 to 3.3.1 ([4ee2531](https://github.com/justintaddei/arp-lookup/commit/4ee25318beb30fe08118549379bd702204791496))
* **deps:** bump actions/setup-node from 2.5.1 to 3.0.0 ([76601c9](https://github.com/justintaddei/arp-lookup/commit/76601c9ec4bec0a8ac6fc3dcafb1519e02431427))
* **deps:** bump actions/setup-node from 3.0.0 to 3.1.0 ([a4974d4](https://github.com/justintaddei/arp-lookup/commit/a4974d42393264fbd18e5a49bfda684aacb869fa))
* **deps:** bump actions/setup-node from 3.1.0 to 3.1.1 ([ce474bd](https://github.com/justintaddei/arp-lookup/commit/ce474bd75ba7d62602d54530696a0666837a90fd))
* **deps:** bump actions/setup-node from 3.1.1 to 3.2.0 ([7df5df0](https://github.com/justintaddei/arp-lookup/commit/7df5df0bafb6e0fb9e083297aa02622121d5203b))
* **deps:** bump actions/setup-node from 3.2.0 to 3.3.0 ([398acd2](https://github.com/justintaddei/arp-lookup/commit/398acd2bd1a9c7f42ae505f615c3064aee82b913))
* **deps:** bump actions/setup-node from 3.3.0 to 3.6.0 ([418bb06](https://github.com/justintaddei/arp-lookup/commit/418bb06e9d9d246ab836d2d5ffdd45249fd1c7dc))
* **deps:** bump dependabot/fetch-metadata from 1.2.1 to 1.3.0 ([63cdf08](https://github.com/justintaddei/arp-lookup/commit/63cdf08b69f2a9683204f02f799da1fb0442759b))
* **deps:** bump dependabot/fetch-metadata from 1.3.0 to 1.3.1 ([7e2f773](https://github.com/justintaddei/arp-lookup/commit/7e2f773dda30794cd55ce8e4c89bb214e444603e))
* **deps:** bump dependabot/fetch-metadata from 1.3.1 to 1.3.2 ([1cf73e8](https://github.com/justintaddei/arp-lookup/commit/1cf73e886965abb603c481d4aec949561aa1a23f))
* **deps:** bump dependabot/fetch-metadata from 1.3.2 to 1.4.0 ([f484d18](https://github.com/justintaddei/arp-lookup/commit/f484d180689bb8d84eeb0085c1be98f7a5e9f72f))
* **deps:** bump minimist from 1.2.5 to 1.2.6 ([69b6441](https://github.com/justintaddei/arp-lookup/commit/69b64411f48f00301ce505d13dce8cb9f124bb04))

## [2.0.0](https://github.com/justintaddei/arp-lookup/compare/v1.3.5...v2.0.0) (2022-02-24)


### ⚠ BREAKING CHANGES

* `is()` renamed to `isType()`
* removed vendor information

### Bug Fixes

* **deps:** bump actions/cache from 2.1.6 to 2.1.7 ([e85d20a](https://github.com/justintaddei/arp-lookup/commit/e85d20a92a5db409ed0a0cf85595907ab0d314de))
* **deps:** bump actions/setup-node from 2.4.0 to 2.5.1 ([8738410](https://github.com/justintaddei/arp-lookup/commit/87384108f6e77301867cdbae6876538aaf5d5e59))
* **deps:** bump dependabot/fetch-metadata from 1.1.1 to 1.2.0 ([7e5d16b](https://github.com/justintaddei/arp-lookup/commit/7e5d16b64cf523449941f4ec469182c6e63ffbfa))
* **deps:** bump dependabot/fetch-metadata from 1.2.0 to 1.2.1 ([6e3d3e6](https://github.com/justintaddei/arp-lookup/commit/6e3d3e6dec33d7fc6f3b464e48a8e1f64ae79d1c))
* **deps:** bump path-parse from 1.0.6 to 1.0.7 ([147fbbf](https://github.com/justintaddei/arp-lookup/commit/147fbbf2dad37077b21ca5aa80d40f6434698134))


* `is()` renamed to `isType()` ([f9e9be0](https://github.com/justintaddei/arp-lookup/commit/f9e9be07302a8df53c372784bdf36cf3685539cf))
* removed vendor information ([e2e66cc](https://github.com/justintaddei/arp-lookup/commit/e2e66cc151eff2bb7e08a34b6714729d1579a3a4))

### [1.3.5](https://github.com/justintaddei/arp-lookup/compare/v1.3.4...v1.3.5) (2021-08-26)


### Bug Fixes

* **deps:** bump actions/setup-node from 2.2.0 to 2.3.0 ([b27a1eb](https://github.com/justintaddei/arp-lookup/commit/b27a1eb307bb0601a35e2f09b387f0ece4dc81fe))

__test__/arp-lookup.test.ts
jest.mock('child_process')

import arp from '../src/index'

test('Arp table is parsed', async done => {
  await expect(arp.getTable()).resolves.toEqual([
    { ip: '192.168.137.255', mac: 'ff:ff:ff:ff:ff:ff', type: 'static', vendor: '' },
    { ip: '224.0.0.22', mac: '01:00:5e:00:00:16', type: 'static', vendor: '' },
    { ip: '224.0.0.251', mac: '01:00:5e:00:00:fb', type: 'static', vendor: '' },
    { ip: '224.0.0.252', mac: '01:00:5e:00:00:fc', type: 'static', vendor: '' },
    { ip: '239.255.255.250', mac: '01:00:5e:7f:ff:fa', type: 'static', vendor: '' },
    { ip: '255.255.255.255', mac: 'ff:ff:ff:ff:ff:ff', type: 'static', vendor: '' },
    { ip: '192.168.2.1', mac: '04:a1:51:1b:12:92', type: 'dynamic', vendor: 'Netgear' },
    { ip: '192.168.2.255', mac: 'ff:ff:ff:ff:ff:ff', type: 'static', vendor: '' },
    { ip: '224.0.0.2', mac: '01:00:5e:00:00:02', type: 'static', vendor: '' },
    { ip: '224.0.0.22', mac: '01:00:5e:00:00:16', type: 'static', vendor: '' },
    { ip: '224.0.0.251', mac: '01:00:5e:00:00:fb', type: 'static', vendor: '' },
    { ip: '224.0.0.252', mac: '01:00:5e:00:00:fc', type: 'static', vendor: '' },
    { ip: '239.255.255.250', mac: '01:00:5e:7f:ff:fa', type: 'static', vendor: '' },
    { ip: '255.255.255.255', mac: 'ff:ff:ff:ff:ff:ff', type: 'static', vendor: '' }
  ])
  done()
})

describe('Arp conversions', () => {
  test('Convert from ip to mac', async done => {
    const mac = await arp.toMAC('192.168.2.1')
    expect(mac).toBe('04:a1:51:1b:12:92')
    done()
  })
  test('Convert from mac to ip', async done => {
    const ip = await arp.toIP('04:a1:51:1b:12:92')
    expect(ip).toBe('192.168.2.1')
    done()
  })
  describe('arp.is()', () => {
    test('static mac', async done => {
      await expect(arp.is('static', '01:00:5e:00:00:02')).resolves.toBe(true)
      done()
    })
    test('static ip', async done => {
      await expect(arp.is('static', '192.168.2.255')).resolves.toBe(true)
      done()
    })
    test('dynamic mac', async done => {
      await expect(arp.is('dynamic', '04:a1:51:1b:12:92')).resolves.toBe(true)
      done()
    })
    test('dynamic ip', async done => {
      await expect(arp.is('dynamic', '192.168.2.1')).resolves.toBe(true)
      done()
    })
    test('undefined address', async done => {
      await expect(arp.is('undefined', '0.0.0.0')).resolves.toBe(true)
      done()
    })
  })
})

describe('Validation works properly', () => {
  describe('MAC addresses', () => {
    test('Throw on invalid mac', async done => {
      await expect(arp.toIP('not:a:mac')).rejects.toThrowError('Invalid MAC')
      done()
    })
    test('is() throws on invalid mac', async done => {
      await expect(arp.is('undefined', 'not:a:mac')).rejects.toThrowError('Invalid address')
      done()
    })
  })
  describe('IP addresses', () => {
    test('Throw on invalid ip', async done => {
      await expect(arp.toMAC('not.an.ip')).rejects.toThrowError('Invalid IP')
      done()
    })
    test('is() throws on invalid ip', async done => {
      await expect(arp.is('undefined', 'not.an.ip')).rejects.toThrowError('Invalid address')
      done()
    })
  })
})


__test__/arp-lookup-lin.test.ts
jest.mock('child_process')

import arp from '../src/index'

describe('On a Linux environment', () => {
  Object.defineProperty(process, 'platform', {
    value: 'linux',
  })
  test('Arp table is parsed', async () => {
    await expect(arp.getTable()).resolves.toMatchSnapshot()
  })
})


__test__/arp-lookup-mac.test.ts
jest.mock('child_process')

import arp from '../src/index'

describe('On a Mac environment', () => {
  Object.defineProperty(process, 'platform', {
    value: 'darwin',
  })
  test('Arp table is parsed', async () => {
    await expect(arp.getTable()).resolves.toMatchSnapshot()
  })
})


__mocks__/child_process.ts
const arpTableWin = `Interface: 192.168.137.1 --- 0x2
  Internet Address      Physical Address      Type      Vendor
  192.168.137.255       ff-ff-ff-ff-ff-ff     static
  224.0.0.22            01-00-5e-00-00-16     static
  224.0.0.251           01-00-5e-00-00-fb     static
  224.0.0.252           01-00-5e-00-00-fc     static
  239.255.255.250       01-00-5e-7f-ff-fa     static
  255.255.255.255       ff-ff-ff-ff-ff-ff     static

Interface: 192.168.2.2 --- 0x9
  Internet Address      Physical Address      Type      Vendor
  192.168.2.1           04-a1-51-1b-12-92     dynamic
  192.168.2.255         ff-ff-ff-ff-ff-ff     static
  224.0.0.2             01-00-5e-00-00-02     static
  224.0.0.22            01-00-5e-00-00-16     static
  224.0.0.251           01-00-5e-00-00-fb     static
  224.0.0.252           01-00-5e-00-00-fc     static
  239.255.255.250       01-00-5e-7f-ff-fa     static
  255.255.255.255       ff-ff-ff-ff-ff-ff     static`

const arpTableMac = `? (192.168.137.255) at ff-ff-ff-ff-ff-ff on en9 ifscope [ethernet]
  ? (224.0.0.22) at 01-00-5e-00-00-16 on en9 ifscope [ethernet]
  ? (224.0.0.251) at 01-00-5e-00-00-fb on en9 ifscope [ethernet]
  ? (224.0.0.252) at 01-00-5e-00-00-fc on en9 ifscope [ethernet]
  ? (239.255.255.250) at 01-00-5e-7f-ff-fa on en9 ifscope [ethernet]
  ? (255.255.255.255) at ff-ff-ff-ff-ff-ff on en9 ifscope [ethernet]
  ? (192.168.2.1) at 04-a1-51-1b-12-92 on en9 ifscope [ethernet]
  ? (192.168.2.255) at ff-ff-ff-ff-ff-ff on en9 ifscope [ethernet]
  ? (224.0.0.2) at 01-00-5e-00-00-02 on en9 ifscope [ethernet]
  ? (224.0.0.22) at 01-00-5e-00-00-16 on en9 ifscope [ethernet]
  ? (224.0.0.251) at 01-00-5e-00-00-fb on en9 ifscope [ethernet]
  ? (224.0.0.252) at 01-00-5e-00-00-fc on en9 ifscope [ethernet]
  ? (239.255.255.250) at 01-00-5e-7f-ff-fa on en9 ifscope [ethernet]
  ? (255.255.255.255) at ff-ff-ff-ff-ff-ff on en9 ifscope [ethernet]`

const arpTableLinux = `? (192.168.0.14) at 1s:21:f3:f1:fe:f2 [ether] on wlan0
? (192.168.0.13) at 01:00:5e:7f:ff:fa [ether] on wlan0
? (192.168.0.3) at ff:ff:ff:ff:ff:ff [ether] on wlan0
? (192.168.0.2) at 01:00:5e:00:00:fb [ether] on wlan0
? (192.168.0.7) at 01:00:5e:00:00:16 [ether] on wlan0
? (192.168.0.6) at 01:00:5e:00:00:02 [ether] on wlan0
? (192.168.0.5) at 01:00:5e:7f:ff:fa [ether] on wlan0`

export function exec(cmd: string, cb: (err: Error | undefined, stdout: string) => void) {
  if (cmd !== 'arp -a' && cmd !== 'arp -an') cb(Error('Unknown command'), 'unknown command')

  const platform = process.platform.substring(0, 3);
  if (platform === 'win') {
    cb(undefined, arpTableWin)
  } else if(platform === 'lin') {
    cb(undefined, arpTableLinux)
  } else {
    cb(undefined, arpTableMac)
  }
}


src/index.ts
import { exec } from 'child_process'
import { isIP } from 'net'
export interface IArpTableRow {
  ip: string
  mac: string
  type: 'static' | 'dynamic' | 'unknown'
}

export type IArpTable = IArpTableRow[]

/**
 * Normalizes a MAC address so that `-` is
 * replaced `:` and is converted to lower case
 *
 * Example: `04-A1-51-1B-12-92` => `04:a1:51:1b:12:92`
 * @param mac The MAC Address to normalize
 */
const normalize = (mac: string): string => mac.replace(/-/g, ':').toLowerCase()

/**
 * Fixes the non compliant MAC addresses returned on Apple systems by adding a leading 0 on parts of the address
 * @param mac The MAC address to FIX
 */
const fixMAC = (mac: string): string =>
  normalize(mac)
    .split(':')
    .map((part) => (part.length === 1 ? '0' + part : part))
    .join(':')

/**
 * Checks if a MAC address is valid
 * @param mac The MAC address to validate
 */
export const isMAC = (mac: string): boolean => /^([0-9A-F]{2}[:-]){5}([0-9A-F]{2})$/i.test(mac)

/**
 * Checks if a MAC address prefix is valid
 * @param prefix The prefix to validate
 */
export const isPrefix = (prefix: string): boolean => /^([0-9A-F]{2}[:-]){2}([0-9A-F]{2})$/i.test(prefix)

/**
 * Retrieves the networks' arp table
 */
export function getTable(): Promise<IArpTable> {
  return new Promise((resolve, reject) => {
    const isWindows = process.platform.substring(0, 3) === 'win'
    const command = isWindows ? 'arp -a' : 'arp -an'

    exec(command, (err, rawArpData) => {
      if (err) {
        reject(err)
        return
      }

      const rows = rawArpData.split('\n')

      const table: IArpTable = []

      for (const row of rows) {
        let ip: string
        let mac: string
        let type: IArpTableRow['type']

        if (isWindows) {
          ;[ip, mac, type] = row.trim().replace(/\s+/g, ' ').split(' ') as [string, string, IArpTableRow['type']]
        } else {
          // Parse the rows as they are returned on unix (Mac or Linux) systems
          const match = /.*\((.*?)\) \w+ (.{0,17}) (?:\[ether]|on)/g.exec(row)
          if (match && match.length === 3) {
            ip = match[1]
            mac = fixMAC(match[2])
            type = 'unknown'
          } else {
            continue
          }
        }

        /*
          If `ip` isn't a valid IP address or `mac` isn't
          a valid MAC address then this is a header
          row (e.g. "Interface: 192.168.137.1 --- 0x2")
          so we can just ignore it.
        */
        if (!isIP(ip) || !isMAC(mac)) continue

        table.push({
          ip,
          mac: normalize(mac),
          type,
        })
      }

      resolve(table)
    })
  })
}

/**
 * Gets the MAC address for the given `ip` address
 * @param ip The IP address
 */
export async function toMAC(ip: string): Promise<string | null> {
  if (!isIP(ip)) throw Error('Invalid IP')

  const arpTable = await getTable()
  const match = arpTable.find((row) => row.ip === ip)

  return match ? match.mac : null
}

/**
 * Gets the IP address for given `mac` address
 * @param mac The MAC address
 */
export async function toIP(mac: string): Promise<string | null> {
  if (!isMAC(mac)) throw Error('Invalid MAC')

  mac = normalize(mac)

  const arpTable = await getTable()

  const match = arpTable.find((row) => row.mac === mac)

  return match ? match.ip : null
}

/**
 * Returns all devices on the network with
 * the specified MAC prefix
 * @param prefix the prefix to search for
 */
export async function fromPrefix(prefix: string): Promise<IArpTableRow[]> {
  if (!isPrefix(prefix)) throw Error('Invalid Prefix')

  const table = await getTable()

  return table.filter((row) => row.mac.startsWith(prefix))
}

export async function isType(type: IArpTableRow['type'] | 'undefined', address: string): Promise<boolean> {
  if (process.platform === 'darwin' && type !== 'unknown' && process.env.NODE_ENV !== 'production')
    console.warn('[arp-lookup] `isType` will always return `false` for types other than "unknown" on darwin systems')

  return type === (await getType(address))
}

export async function getType(address: string): Promise<string> {
  if (!isIP(address) && !isMAC(address)) throw Error('Invalid address')

  if (process.platform === 'darwin' && process.env.NODE_ENV !== 'production')
    console.warn('[arp-lookup] `getType` will always return "unknown" on darwin systems')

  if (isMAC(address)) address = normalize(address)

  const recordedType = (await getTable()).find((row) => row.ip === address || row.mac === address)?.type ?? 'undefined'

  return recordedType
}

export default {
  fromPrefix,
  getTable,
  getType,
  isType,
  isIP,
  isMAC,
  isPrefix,
  toIP,
  toMAC,
}
